types/styled-components/index.d.ts

```diff
@@ -132,101 +254,135 @@ export interface BaseThemedCssFunction<T> {
 
-export interface ThemeConsumerProps<T> {
-    children(theme: T): React.ReactNode;
+// tslint:disable-next-line:no-empty-interface
+export interface DefaultTheme {}
+
+export interface ThemeProviderProps<T extends object> {
+    children?: React.ReactChild; // only one child is allowed, goes through React.Children.only
+    theme: T | ((theme: T) => T);
 }
-export type BaseThemeConsumerComponent<T> = React.ComponentClass<ThemeConsumerProps<T>>;
-export type ThemeConsumerComponent<T> = BaseThemeConsumerComponent<Extract<keyof T, string> extends never ? any : T>;
-export const ThemeConsumer: ThemeConsumerComponent<object>;
+export type BaseThemeProviderComponent<T extends object> = React.ComponentClass<
+    ThemeProviderProps<T>
+>;
+export type ThemeProviderComponent<
+    T extends object
+> = BaseThemeProviderComponent<Extract<keyof T, string> extends never ? any : T>;
+export const ThemeProvider: ThemeProviderComponent<DefaultTheme>;
+// NOTE: this technically starts as undefined
+// Also, this cannot be DefaultTheme as it breaks TypedComponents' assignability
+export const ThemeContext: React.Context<any>;
+export const ThemeConsumer: typeof ThemeContext['Consumer'];
 
 export interface Keyframes {
     getName(): string;
 }
 
+export function keyframes(cssObject: CSSObject): Keyframes;
 export function keyframes(
     strings: TemplateStringsArray,
     ...interpolations: SimpleInterpolation[]
 ): Keyframes;
 
+export function createGlobalStyle(
+    cssObject: CSSObject
+): GlobalStyleComponent<{}, DefaultTheme>;
 export function createGlobalStyle<P = {}>(
     strings: TemplateStringsArray,
     ...interpolations: Array<Interpolation<ThemedStyledProps<P, DefaultTheme>>>
-): GlobalStyleClass<P, DefaultTheme>;
+): GlobalStyleComponent<P, DefaultTheme>;
+export function createGlobalStyle<P = {}>(
+    func: InterpolationFunction<ThemedStyledProps<P, DefaultTheme>>
+): GlobalStyleComponent<P, DefaultTheme>;
 
 export function isStyledComponent(
-    target: any,
-): target is StyledComponentClass<{}, {}>;
-
-export const ThemeProvider: ThemeProviderComponent<object>;
-
-interface StylesheetComponentProps {
-    sheet: ServerStyleSheet;
-}
-
-interface StyleSheetManagerProps {
-    sheet?: StyleSheet;
-    target?: Node;
-}
-
-export class StyleSheetManager extends React.Component<
-    StyleSheetManagerProps
-> {}
+    target: any
+): target is StyledComponent<any, any>;
 
 export class ServerStyleSheet {
     collectStyles(
-        tree: React.ReactNode,
-    ): React.ReactElement<StylesheetComponentProps>;
+        tree: React.ReactNode
+    ): React.ReactElement<{ sheet: ServerStyleSheet }>;
 
     getStyleTags(): string;
     getStyleElement(): Array<React.ReactElement<{}>>;
     interleaveWithNodeStream(
-        readableStream: NodeJS.ReadableStream,
+        readableStream: NodeJS.ReadableStream
     ): NodeJS.ReadableStream;
-    instance: StyleSheet;
+    readonly instance: this;
```

Comment:
The `StyleSheet` in the source code is not the DOM `StyleSheet` which this was referring to but a private class that is not exported so I didn't define it here.

At any rate, the `instance` at runtime will be a `ServerStyleSheet` instance anyway so I just made it be `this` typed.